back/engines/commercial/report_builder/app/services/report_builder/queries/visitors.rb

```diff
@@ -0,0 +1,145 @@
+module ReportBuilder
+  class Queries::Visitors < ReportBuilder::Queries::Base
+    def run_query(
+      start_at: nil,
+      end_at: nil,
+      project_id: nil,
+      resolution: 'month',
+      exclude_roles: nil,
+      compare_start_at: nil,
+      compare_end_at: nil,
+      **_other_props
+    )
+      validate_resolution(resolution)
+      validate_roles(exclude_roles)
```

Comment:
I have been using these validate-utilities to avoid SQL injections in other queries as well. Not sure if there is a more conventional way of doing this?